llvm/lib/CodeGen/SelectionDAG/DAGCombiner.cpp

```diff
@@ -23497,6 +23497,29 @@ SDValue DAGCombiner::visitBUILD_VECTOR(SDNode *N) {
       return DAG.getNode(ISD::SPLAT_VECTOR, SDLoc(N), VT, V);
     }
 
+  // build_vector(build_pair(x,y)) -> bitcast(build_pair(x,y))
+  if (N->getNumOperands() == 1 &&
+      N->getOperand(0).getOpcode() == ISD::BUILD_PAIR)
+    return DAG.getBitcast(VT, N->getOperand(0));
+
```

Comment:
I'm surprised you get away without any legality checks

Reply:
BUILD_PAIR nodes get legalized away pretty early